plugins/aws/providers.go

```diff
@@ -0,0 +1,142 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/1Password/shell-plugins/sdk"
+	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+	confighelpers "github.com/99designs/aws-vault/v7/vault"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+type assumeRoleProvider struct {
+	confighelpers.AssumeRoleProvider
+	stsCacheWriter
+}
+
+func (p assumeRoleProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.AssumeRoleProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewAssumeRoleProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN)
+	if in.Cache.Has(roleCacheKey) {
+		return NewStsCacheProvider(roleCacheKey, in.Cache)
+	}
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
```

Comment:
What if there is a serial but no token? Or vice-versa? I think we'd want to return an error in this case to make it easier to debug for the user. As when they have either set, they clearly want to use MFA, so I believe falling back to not using MFA would make things harder for these users.

Reply:
Yes, that's right. Added that extra check now. thanks!